File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/honorarios/TotalizadorDeHonorario.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.honorarios;

import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.Total;
import br.jus.trt8.pjecalc.negocio.constantes.TipoCobrancaReclamanteEnum;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.Calculo;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.honorarios.Honorario;
import java.io.Serializable;
import java.math.BigDecimal;

public class TotalizadorDeHonorario
implements Serializable {
    private static final long serialVersionUID = 1L;
    private Calculo calculo;
    private boolean isCalculado;
    private Total honorariosDevidosPeloReclamante;
    private Total honorariosDevidosPeloReclamado;

    public TotalizadorDeHonorario(Calculo calculo) {
        this.calculo = calculo;
        this.reset();
    }

    public void reset() {
        this.isCalculado = false;
    }

    private TotalizadorDeHonorario calcular() {
        if (!this.isCalculado) {
            this.honorariosDevidosPeloReclamante = Total.newInstance(true);
            this.honorariosDevidosPeloReclamado = Total.newInstance(true);
            for (Honorario honorario : this.calculo.getHonorariosDoCalculo()) {
                if (!Utils.naoNulo(honorario.getValorTotal())) continue;
                switch (honorario.getTipoDeDevedor()) {
                    case RECLAMANTE: {
                        boolean isCobrarDoReclamante = honorario.getTipoCobrancaReclamante() == TipoCobrancaReclamanteEnum.COBRAR;
                        this.honorariosDevidosPeloReclamante.acumular(isCobrarDoReclamante ? BigDecimal.ZERO : honorario.getValorTotal());
                        break;
                    }
                    case RECLAMADO: {
                        this.honorariosDevidosPeloReclamado.acumular(honorario.getValorTotal());
                    }
                }
            }
            this.isCalculado = true;
        }
        return this;
    }

    public BigDecimal getTotalDevidoPeloReclamante() {
        return this.calcular().honorariosDevidosPeloReclamante.getValor();
    }

    public BigDecimal getTotalDevidoPeloReclamado() {
        return this.calcular().honorariosDevidosPeloReclamado.getValor();
    }
}

File: src/lib/pjecalc/core/dominio/calculo/honorarios/totalizador-de-honorario.ts
/**
 * PJe-Calc v2.15.1 — TotalizadorDeHonorario
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.honorarios.TotalizadorDeHonorario
 *
 * Ref Java: pjecalc-fonte/.../calculo/honorarios/TotalizadorDeHonorario.java
 *
 * Soma honorários por devedor (reclamante/reclamado). Honorários cobrados
 * diretamente do reclamante (tipoCobranca=COBRAR) não entram na conta do
 * reclamante — pois nesse caso vão a pagar separadamente.
 * Cache lazy: `reset()` invalida.
 */
import Decimal from 'decimal.js';
import type { Calculo } from '../calculo';
import { TipoCobrancaReclamanteEnum, TipoDeDevedorDoHonorarioEnum } from '../../../constantes/enums';
import type { Honorario } from './honorario';

const ZERO = new Decimal(0);

class Total {
  private valor: Decimal = ZERO;
  acumular(v: Decimal | null): void { if (v !== null) this.valor = this.valor.plus(v); }
  getValor(): Decimal { return this.valor; }
}

export class TotalizadorDeHonorario {
  private calculo: Calculo;
  private isCalculado = false;
  private honorariosDevidosPeloReclamante = new Total();
  private honorariosDevidosPeloReclamado = new Total();

  constructor(calculo: Calculo) {
    this.calculo = calculo;
  }

  reset(): void { this.isCalculado = false; }

  private calcular(): this {
    if (this.isCalculado) return this;
    this.honorariosDevidosPeloReclamante = new Total();
    this.honorariosDevidosPeloReclamado = new Total();
    // ducktype — Calculo.getHonorariosDoCalculo() não está tipado ainda (Fase 10).
    const calcExt = this.calculo as unknown as { getHonorariosDoCalculo?(): Iterable<Honorario> };
    for (const honorario of calcExt.getHonorariosDoCalculo?.() ?? []) {
      const total = honorario.getValorTotal();
      if (total === null) continue;
      switch (honorario.getTipoDeDevedor()) {
        case TipoDeDevedorDoHonorarioEnum.RECLAMANTE: {
          const isCobrarDoReclamante = honorario.getTipoCobrancaReclamante() === TipoCobrancaReclamanteEnum.COBRAR;
          this.honorariosDevidosPeloReclamante.acumular(isCobrarDoReclamante ? ZERO : total);
          break;
        }
        case TipoDeDevedorDoHonorarioEnum.RECLAMADO:
          this.honorariosDevidosPeloReclamado.acumular(total);
          break;
      }
    }
    this.isCalculado = true;
    return this;
  }

  getTotalDevidoPeloReclamante(): Decimal {
    return this.calcular().honorariosDevidosPeloReclamante.getValor();
  }

  getTotalDevidoPeloReclamado(): Decimal {
    return this.calcular().honorariosDevidosPeloReclamado.getValor();
  }
}
